### PR TITLE
Migrate the agent-server communication from RMI to WebSocket

### PR DESCRIPTION
**Pull request by dim4o**
Add a WebSocket layer to the Agent
Remove the RMI imports and exceptions and other trivial additions and removals
Apply the WebSocket layer to the rest part of the Agent's project

### CODE REVIEW

File: src/main/java/com/musala/atmosphere/agent/Agent.java [hunk 1]
```diff
@@ -79,7 +77,6 @@ public Agent(int agentRmiPort) {
         this.agentRmiPort = agentRmiPort;
```
Comment (Perseverance) [inline src/main/java/com/musala/atmosphere/agent/Agent.java]: Fix RMI namings and javadocs

File: src/main/java/com/musala/atmosphere/agent/AgentManager.java [hunk 2]
```diff
@@ -68,10 +52,8 @@
      *        Port, which will be used for the RMI Registry
      * @param fileRecycler
      *        {@link FileRecycler FileRecycler} object
-     * @throws RemoteException
-     *         - required when implementing {@link UnicastRemoteObject}
      */
-    public AgentManager(int rmiPort, FileRecycler fileRecycler) throws RemoteException {
+    public AgentManager(int rmiPort, FileRecycler fileRecycler) {
```
Comment (Perseverance) [inline src/main/java/com/musala/atmosphere/agent/AgentManager.java]: remove rmi Namings

File: src/main/java/com/musala/atmosphere/agent/AgentManager.java [hunk 3]
```diff
@@ -162,32 +129,47 @@ public String createAndStartEmulator(EmulatorParameters parameters) throws Remot
...
+    /**
+     * Registers the server for events, related to changes in the devices on the agent. Should be called after the
+     * server has published it's IAgentEventSender.
+     *
+     * @param serverIPAddress
+     *        The server's RMI IP address.
+     * @param serverRmiPort
+     *        Port on which the RMI registry is opened.
+     */
+    public void registerServer(String serverIPAddress, int serverRmiPort) {
```
Comment (Perseverance) [inline src/main/java/com/musala/atmosphere/agent/AgentManager.java]: We should revise this code and see if it is needed in this way

File: src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java [hunk 4]
```diff
@@ -156,11 +120,7 @@ private void updateDevice(IDevice device) {
      * @return <code>true</code> if the device is registered and <code>false</code> otherwise.
      */
     private boolean isDeviceRegistered(IDevice device) {
-        try {
-            String deviceRmiId = device.getSerialNumber();
-            return deviceManager.getAllDeviceRmiIdentifiers().contains(deviceRmiId);
-        } catch (RemoteException e) {
-            return false;
-        }
+        String deviceRmiId = device.getSerialNumber();
```
Comment (Perseverance) [inline src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java]: I don not think this variable name is correct

File: src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java [hunk 5]
```diff
@@ -156,11 +120,7 @@ private void updateDevice(IDevice device) {
...
+        return deviceManager.getAllDeviceRmiIdentifiers().contains(deviceRmiId);
```
Comment (Perseverance) [inline src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java]: Is this method correct?

File: src/main/java/com/musala/atmosphere/agent/DeviceManager.java [hunk 6]
```diff
@@ -96,11 +91,16 @@
...
+     * Maps a device serial number to a {@link IWrapDevice} device wrapper
+     */
+    private static Map<String, IWrapDevice> deviceSerialToDeviceWrapper = new HashMap<>();
+
     public DeviceManager() {
     }
 
-    public DeviceManager(int rmiPort, FileRecycler fileRecycler) throws RemoteException {
-        if(AgentPropertiesLoader.hasFtpServer() && ftpFileTransferService == null) {
+    public DeviceManager(int rmiPort, FileRecycler fileRecycler) {
```
Comment (Perseverance) [inline src/main/java/com/musala/atmosphere/agent/DeviceManager.java]: Change the namings and javadocs containing RMI

File: src/main/java/com/musala/atmosphere/agent/Agent.java (hunk 1 above)
Comment (dim4o) [inline src/main/java/com/musala/atmosphere/agent/Agent.java]: done

File: src/main/java/com/musala/atmosphere/agent/AgentManager.java (hunk 2 above)
Comment (dim4o) [inline src/main/java/com/musala/atmosphere/agent/AgentManager.java]: done

File: src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java (hunk 4 above)
Comment (dim4o) [inline src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java]: done

File: src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java (hunk 5 above)
Comment (dim4o) [inline src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java]: done

File: src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java [hunk 7]
```diff
@@ -28,24 +26,18 @@ public DeviceChangeListener() {
...
      *
-     * @param serverIPAddress
-     *        - IP address of the server's RMI registry
-     * @param serverRmiPort
-     *        - Port on which the server's RMI registry is opened
+     * @param isServerSet
+     *        - whether the server is set
      */
-    public DeviceChangeListener(String serverIPAddress, int serverRmiPort) {
+    public DeviceChangeListener(boolean isServerSet) {
```
Comment (Perseverance) [inline src/main/java/com/musala/atmosphere/agent/DeviceChangeListener.java]: Change the name to something more meaningful like isConnectedToServer or isServerConnected

File: src/main/java/com/musala/atmosphere/agent/websocket/AgentDispatcher.java [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+    /**
+     * Closes the connection with the server.
+     */
+    public void close() {
+        if (session != null && session.isOpen()) {
+            try {
+                session.close();
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (bstrandjev) [inline src/main/java/com/musala/atmosphere/agent/websocket/AgentDispatcher.java]: Logger.error

File: src/main/java/com/musala/atmosphere/agent/AgentManager.java [hunk 9]
```diff
@@ -162,37 +127,52 @@ public String createAndStartEmulator(EmulatorParameters parameters) throws Remot
...
+    /**
+     * Registers the server for events, related to changes in the devices on the agent. Should be called after the
+     * server has published it's IAgentEventSender.
+     *
+     * @param serverIpAddress
+     *        The server's IP address.
+     * @param serverPort
+     *        Port on which the WebSocket connection is opened.
+     */
+    public void registerServer(String serverIpAddress, int serverPort) {
```
Comment (bstrandjev) [inline src/main/java/com/musala/atmosphere/agent/AgentManager.java]: this is confusing. Why do you pass in parameters that you do not use anywhere but in logging?

File: src/main/java/com/musala/atmosphere/agent/DeviceChangeHandler.java [hunk 10]
```diff
@@ -144,7 +130,7 @@ public void run() {
             if (publishId != null && !publishId.isEmpty()) {
                 try {
                     onDeviceListChanged(publishId, false /* device disconnected */);
-                } catch (CommandFailedException | NotBoundException e) {
+                } catch (CommandFailedException e) {
```
Comment (bstrandjev) [inline src/main/java/com/musala/atmosphere/agent/DeviceChangeHandler.java]: log error here, do not just silently ignore it

File: src/main/java/com/musala/atmosphere/agent/DeviceManager.java [hunk 11]
```diff
@@ -163,7 +159,7 @@ public void run() {
                 });
             }
 
-            ChromeDriverManager.getInstance().setup();
+            ChromeDriverManager.getInstance().setup("2.27");
```
Comment (bstrandjev) [inline src/main/java/com/musala/atmosphere/agent/DeviceManager.java]: what is this magic?

File: src/main/java/com/musala/atmosphere/agent/websocket/AgentDispatcher.java (hunk 8 above)
Comment (dim4o) [inline src/main/java/com/musala/atmosphere/agent/websocket/AgentDispatcher.java]: Done

File: src/main/java/com/musala/atmosphere/agent/AgentManager.java (hunk 9 above)
Comment (dim4o) [inline src/main/java/com/musala/atmosphere/agent/AgentManager.java]: Changed

File: src/main/java/com/musala/atmosphere/agent/DeviceChangeHandler.java (hunk 10 above)
Comment (dim4o) [inline src/main/java/com/musala/atmosphere/agent/DeviceChangeHandler.java]: Done

File: src/main/java/com/musala/atmosphere/agent/DeviceManager.java (hunk 11 above)
Comment (dim4o) [inline src/main/java/com/musala/atmosphere/agent/DeviceManager.java]: It was a "dirty" fix on a failing Travis build.  Removed.

**Pull request merged**